File: scripts/fetch-hangman-nba-players.mjs
/**
 * Fetches all NBA roster players for the current ESPN season (e.g. 2025-26)
 * and writes data/minigames/nba-players-2025-26.json
 *
 * Each player includes: photoUrl, height, nationality, division, age.
 * Draft year is not exposed on this roster endpoint (use a separate source if needed).
 *
 * Run: node scripts/fetch-hangman-nba-players.mjs (Node 18+)
 * Or:  python3 scripts/fetch_hangman_nba_players.py
 * Requires: network
 */
import { writeFileSync, mkdirSync } from "fs"
import { dirname, join } from "path"
import { fileURLToPath } from "url"

const __dirname = dirname(fileURLToPath(import.meta.url))
const OUT = join(__dirname, "..", "data", "minigames", "nba-players-2025-26.json")

const EAST = new Set([
  "ATL",
  "BOS",
  "BKN",
  "CHA",
  "CHI",
  "CLE",
  "DET",
  "IND",
  "MIA",
  "MIL",
  "NYK",
  "ORL",
  "PHI",
  "TOR",
  "WAS",
])

/** ESPN team abbreviations sometimes differ from the NBA / internal keys above. */
const TEAM_ABBR_ALIASES = {
  NY: "NYK",
  GS: "GSW",
  UTAH: "UTA",
  NO: "NOP",
  SA: "SAS",
  WSH: "WAS",
}

function normalizeTeamAbbr(abbr) {
  if (abbr == null || typeof abbr !== "string") return abbr
  const u = abbr.trim().toUpperCase()
  return TEAM_ABBR_ALIASES[u] ?? u
}

/** NBA division for each franchise (current alignment). */
const TEAM_DIVISION = {
  ATL: "Southeast",
  BOS: "Atlantic",
  BKN: "Atlantic",
  CHA: "Southeast",
  CHI: "Central",
  CLE: "Central",
  DAL: "Southwest",
  DEN: "Northwest",
  DET: "Central",
  GSW: "Pacific",
  HOU: "Southwest",
  IND: "Central",
  LAC: "Pacific",
  LAL: "Pacific",
  MEM: "Southwest",
  MIA: "Southeast",
  MIL: "Central",
  MIN: "Northwest",
  NOP: "Southwest",
  NYK: "Atlantic",
  OKC: "Northwest",
  ORL: "Southeast",
  PHI: "Atlantic",
  PHX: "Pacific",
  POR: "Northwest",
  SAC: "Pacific",
  SAS: "Southwest",
  TOR: "Atlantic",
  UTA: "Northwest",
  WAS: "Southeast",
}

function divisionForAbbr(abbr) {
  const a = normalizeTeamAbbr(abbr)
  return TEAM_DIVISION[a] != null ? TEAM_DIVISION[a] : "Unknown"
}

async function fetchJson(url) {
  const res = await fetch(url, {
    headers: {
      Accept: "application/json",
    },
  })
  if (!res.ok) throw new Error(`${url} -> ${res.status}`)
  return res.json()
}

function conferenceForAbbr(abbr) {
  const a = normalizeTeamAbbr(abbr)
  return EAST.has(a) ? "East" : "West"
}

function main() {
  return (async () => {
    const teamsData = await fetchJson(
      "https://site.api.espn.com/apis/site/v2/sports/basketball/nba/teams?limit=100"
    )
    const teams = teamsData.sports?.[0]?.leagues?.[0]?.teams ?? []
    const players = []

    for (const { team } of teams) {
      const id = team?.id
      const abbrRaw = team?.abbreviation
      const city = team?.location ?? ""
      const nickname = team?.name ?? ""
      const teamName = [city, nickname].filter(Boolean).join(" ").trim() || team?.displayName
      if (!id || !abbrRaw) continue

      const abbr = normalizeTeamAbbr(abbrRaw)
      const roster = await fetchJson(
        `https://site.api.espn.com/apis/site/v2/sports/basketball/nba/teams/${id}/roster`
      )
      const athletes = roster.athletes ?? []
      const conf = conferenceForAbbr(abbr)
      const div = divisionForAbbr(abbr)

      for (const a of athletes) {
        const fullName = a.fullName || a.displayName
        if (!fullName || typeof fullName !== "string") continue
        const pos =
          a.position?.displayName ||
          a.position?.abbreviation ||
          a.position?.type ||
          "Unknown"
        const photoHref =
          typeof a.headshot?.href === "string" && a.headshot.href.length > 0
            ? a.headshot.href
            : null
        const height =
          typeof a.displayHeight === "string" && a.displayHeight.length > 0
            ? a.displayHeight
            : null
        const nationality =
          typeof a.birthPlace?.country === "string" &&
          a.birthPlace.country.length > 0
            ? a.birthPlace.country
            : null
        const age =
          typeof a.age === "number" && Number.isFinite(a.age) ? a.age : null

        const jerseyRaw = a.jersey
        const jerseyNumber =
          jerseyRaw !== undefined &&
          jerseyRaw !== null &&
          String(jerseyRaw).trim().length > 0
            ? String(jerseyRaw).trim()
            : null

        players.push({
          id: `espn-${a.id}`,
          displayName: fullName.trim(),
          team: teamName,
          teamAbbr: abbr,
          conference: conf,
          division: div,
          position: String(pos),
          photoUrl: photoHref,
          height,
          nationality,
          age,
          jerseyNumber,
        })
      }
    }

    const byName = new Map()
    for (const p of players) {
      const key = p.displayName.toLowerCase()
      if (!byName.has(key)) byName.set(key, p)
    }
    const unique = [...byName.values()]

    mkdirSync(dirname(OUT), { recursive: true })
    writeFileSync(
      OUT,
      JSON.stringify(
        {
          seasonLabel: "2025-26",
          source: "ESPN roster API (run scripts/fetch-hangman-nba-players.mjs)",
          updatedAt: new Date().toISOString(),
          players: unique,
        },
        null,
        2
      ),
      "utf8"
    )
    console.log(`Wrote ${unique.length} players to ${OUT}`)
  })()
}

main().catch((e) => {
  console.error(e)
  process.exit(1)
})

File: scripts/fetch_hangman_nba_players.py
#!/usr/bin/env python3
"""
Fetches NBA roster players from ESPN (same data as fetch-hangman-nba-players.mjs)
including photoUrl, height, nationality, division, age.

Run: python3 scripts/fetch_hangman_nba_players.py
Requires: Python 3.9+, network
"""
from __future__ import annotations

import json
import ssl
import urllib.request
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "data" / "minigames" / "nba-players-2025-26.json"

EAST = {
    "ATL", "BOS", "BKN", "CHA", "CHI", "CLE", "DET", "IND", "MIA", "MIL",
    "NYK", "ORL", "PHI", "TOR", "WAS",
}

# ESPN team abbreviations sometimes differ from the NBA keys in EAST / TEAM_DIVISION.
ABBR_ALIASES = {
    "NY": "NYK",
    "GS": "GSW",
    "UTAH": "UTA",
    "NO": "NOP",
    "SA": "SAS",
    "WSH": "WAS",
}


def normalize_team_abbr(abbr: str) -> str:
    if not abbr or not isinstance(abbr, str):
        return abbr
    u = abbr.strip().upper()
    return ABBR_ALIASES.get(u, u)


TEAM_DIVISION = {
    "ATL": "Southeast", "BOS": "Atlantic", "BKN": "Atlantic", "CHA": "Southeast",
    "CHI": "Central", "CLE": "Central", "DAL": "Southwest", "DEN": "Northwest",
    "DET": "Central", "GSW": "Pacific", "HOU": "Southwest", "IND": "Central",
    "LAC": "Pacific", "LAL": "Pacific", "MEM": "Southwest", "MIA": "Southeast",
    "MIL": "Central", "MIN": "Northwest", "NOP": "Southwest", "NYK": "Atlantic",
    "OKC": "Northwest", "ORL": "Southeast", "PHI": "Atlantic", "PHX": "Pacific",
    "POR": "Northwest", "SAC": "Pacific", "SAS": "Southwest", "TOR": "Atlantic",
    "UTA": "Northwest", "WAS": "Southeast",
}


def fetch_json(url: str) -> dict:
    ctx = ssl.create_default_context()
    req = urllib.request.Request(
        url,
        headers={"Accept": "application/json"},
    )
    with urllib.request.urlopen(req, context=ctx, timeout=60) as r:
        return json.loads(r.read().decode())


def conference_for_abbr(abbr: str) -> str:
    a = normalize_team_abbr(abbr)
    return "East" if a in EAST else "West"


def division_for_abbr(abbr: str) -> str:
    a = normalize_team_abbr(abbr)
    return TEAM_DIVISION.get(a, "Unknown")


def main() -> None:
    teams_data = fetch_json(
        "https://site.api.espn.com/apis/site/v2/sports/basketball/nba/teams?limit=100"
    )
    teams = teams_data.get("sports", [{}])[0].get("leagues", [{}])[0].get("teams", [])
    players: list[dict] = []

    for entry in teams:
        team = entry.get("team") or {}
        tid = team.get("id")
        abbr_raw = team.get("abbreviation")
        city = team.get("location") or ""
        nickname = team.get("name") or ""
        parts = [city, nickname]
        team_name = " ".join(p for p in parts if p).strip() or team.get("displayName")
        if not tid or not abbr_raw:
            continue

        abbr = normalize_team_abbr(abbr_raw)
        roster = fetch_json(
            f"https://site.api.espn.com/apis/site/v2/sports/basketball/nba/teams/{tid}/roster"
        )
        athletes = roster.get("athletes") or []
        conf = conference_for_abbr(abbr)
        div = division_for_abbr(abbr)

        for a in athletes:
            full_name = a.get("fullName") or a.get("displayName")
            if not full_name or not isinstance(full_name, str):
                continue
            pos = (
                (a.get("position") or {}).get("displayName")
                or (a.get("position") or {}).get("abbreviation")
                or (a.get("position") or {}).get("type")
                or "Unknown"
            )
            hs = a.get("headshot") or {}
            href = hs.get("href") if isinstance(hs.get("href"), str) else None
            photo = href if href else None

            dh = a.get("displayHeight")
            height = dh if isinstance(dh, str) and dh else None

            bp = a.get("birthPlace") or {}
            nat = bp.get("country") if isinstance(bp.get("country"), str) else None

            age_val = a.get("age")
            age = age_val if isinstance(age_val, (int, float)) and age_val == age_val else None
            if age is not None:
                age = int(age)

            j = a.get("jersey")
            jersey_number = None
            if j is not None and str(j).strip() != "":
                jersey_number = str(j).strip()

            players.append(
                {
                    "id": f"espn-{a.get('id')}",
                    "displayName": full_name.strip(),
                    "team": team_name,
                    "teamAbbr": abbr,
                    "conference": conf,
                    "division": div,
                    "position": str(pos),
                    "photoUrl": photo,
                    "height": height,
                    "nationality": nat,
                    "age": age,
                    "jerseyNumber": jersey_number,
                }
            )

    by_name: dict[str, dict] = {}
    for p in players:
        key = p["displayName"].lower()
        if key not in by_name:
            by_name[key] = p
    unique = list(by_name.values())

    OUT.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "seasonLabel": "2025-26",
        "source": "ESPN roster API (run scripts/fetch_hangman_nba_players.py or fetch-hangman-nba-players.mjs)",
        "updatedAt": datetime.now(timezone.utc)
        .replace(microsecond=0)
        .isoformat()
        .replace("+00:00", "Z"),
        "players": unique,
    }
    OUT.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    print(f"Wrote {len(unique)} players to {OUT}")


if __name__ == "__main__":
    main()
